### PR TITLE
Implement synchronous open method

### PR DESCRIPTION
**Pull request by lachiet**
### Description & motivation

Opening the widget experience relies on being triggered by a click event. This update changes the init functionality to wait until we've received a nonce before allowing the user to open.

- Make open() function synchronous which removes send app params when no nonce is present.
- Remove any queued open options since they could be opened outside of a click event.

_Key changes_

Instead of using "onReady" we'll follow the callback format of `(err, res) => void` to ensure the instance isn't available until we've initialized.
```ts
let onrampInstance: CBPayInstanceType | undefined;
initOnRamp(options, (err, instance) => {
  if (err) {
    // CB Pay unavailable
  }
  onrampInstance = instance;
});
```

### Testing & documentation

Proposed readme update: https://github.com/coinbase/cbpay-js/blob/0e553aab4bc8369203571ce80629fb42394a7e57/README.md

### CODE REVIEW

File: src/utils/CoinbasePixel.ts [hunk 1]
```diff
@@ -32,7 +37,7 @@ export type CoinbasePixelConstructorParams = {
   host?: string;
   appId: string;
   appParams: JsonObject;
-  onReady?: (error?: Error) => void;
+  onReady?: (error?: OnReadyError) => void;
```
Comment (zack-winchell-cb) [inline src/utils/CoinbasePixel.ts:40]: @lachiet Should this be made required at this point?

File: src/utils/CoinbasePixel.ts [hunk 2]
```diff
@@ -117,71 +120,83 @@ export class CoinbasePixel {
...
+          });
+        },
+      );
+    } else if (experience === 'new_tab' && window.chrome?.tabs?.create) {
+      void window.chrome.tabs.create({ url });
+    } else {
+      openWindow(url, experience);
+    }
+
+    // Add an event listener for when the widget opens so we can set a new nonce without invalidating the current nonce
```
Comment (zack-winchell-cb) [inline src/utils/CoinbasePixel.ts:194]: @lachiet Can we get a comment with a bit more context on why we need two nonces at once?

File: src/utils/CoinbasePixel.ts (hunk 1 above)
Comment (lachiet) [reply src/utils/CoinbasePixel.ts]: Yep makes sense.

File: src/utils/CoinbasePixel.ts [hunk 3]
```diff
@@ -231,47 +236,35 @@ export class CoinbasePixel {
...
-  private runQueuedOpenExperience = () => {
-    if (this.queuedOpenOptions) {
-      this.log('Running queued experience');
-      const options = { ...this.queuedOpenOptions };
-      this.queuedOpenOptions = undefined;
-      this.openExperience(options);
+    // If a fallback option is provided we only want to provide a warning since we can still attempt to open the widget
+    if (this.onFallbackOpen) {
+      if (this.debug) {
+        console.warn('Failed to load CB Pay pixel. Falling back to opening in new tab.');
```
Comment (zack-winchell-cb) [inline src/utils/CoinbasePixel.ts:244]: nit: Still good to warn if `!this.onFallbackOpen`.
```suggestion
    if (this.debug) {
      console.warn('Failed to load CB Pay pixel.' + this.onFallbackOpen ? ' Falling back to opening in new tab.' : '');
    }

    if (this.onFallbackOpen) {
```

File: src/utils/CoinbasePixel.ts [hunk 4]
```diff
@@ -191,7 +186,9 @@ export class CoinbasePixel {
       openWindow(url, experience);
     }
 
-    // Add an event listener for when the widget opens so we can set a new nonce without invalidating the current nonce
+    // For users who exit the experience and want to re-enter, we need a fresh nonce to use.
+    // Additionally, if we trigger sendAppParams to early we'll invalidate the nonce they're opening in this current attempt.
```
Comment (zack-winchell-cb) [inline src/utils/CoinbasePixel.ts:190]:
```suggestion
    // Additionally, if we trigger sendAppParams too early we'll invalidate the nonce they're opening in this current attempt.
```

**Review by zack-winchell-cb — APPROVED** [review]

File: src/utils/CoinbasePixel.ts (hunk 3 above)
Comment (lachiet) [reply src/utils/CoinbasePixel.ts:244]: Added an debug error log.

**Review by victor-magalhaes-cb — DISMISSED** [review]

**Review by qodesmith — DISMISSED** [review]

**Review by zack-winchell-cb — APPROVED** [review]

**Review by zack-winchell-cb — APPROVED** [review]

**Review by victor-magalhaes-cb — APPROVED** [review]

**Review by qodesmith — APPROVED** [review]

**Pull request merged**